pkg/operator/terminationobserver/storage.go

```diff
@@ -0,0 +1,147 @@
+package terminationobserver
+
+import (
+	"errors"
+	"sort"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+// The store interface implementation keeps the in-memory state for each API server.
+// The state consists of current creationTimestamp and previous creationTimestamp as well as termination
+// events recorded by termination observer.
+type Store interface {
+	// Get return single API server state
+	Get(name string) (*APIServerState, error)
+
+	// GetEventWithReason return termination event with given reason or nil if such event not found.
+	// This is used in prometheus metrics collector.
+	GetEventWithReason(name, reason string) *corev1.Event
+
+	// ListNames return list of API server names (hostnames)
+	// Used by prometheus metrics collector
+	ListNames() []string
+
+	// Create add new API server state
+	Create(name string, state APIServerState)
+
+	// Update updates the existing API server state.
+	Update(name string, state APIServerState) error
+
+	// UpdateEvents add new events into existing API server state
+	UpdateEvents(name string, event *corev1.Event) error
+}
+
+type memoryStorage struct {
+	sync.RWMutex
+
+	state map[string]APIServerState
+}
+
+type APIServerState struct {
+	// currentCreatedTimestamp is creation timestamp of currently running API server
+	// this API server does not have any termination events, those events are only valid for
+	// "old" API server.
+	currentCreatedTimestamp time.Time
```

Comment:
we also don't postfix other fields with their types. Why timestamps?

Reply:
because this store the 'creationTimestamp' field and I liked to have this matching the original field (or close to original field) if possible.